### PR TITLE
Validate input values when updating IDP claims config

### PR DESCRIPTION
**Pull request by somindatommy**
Resolves https://github.com/wso2/product-is/issues/9702

### Approach
Validate the `userIdClaim` and `roleClaim` with existing role mappings and validate the claim URIs.

### Tests
Following test scenarios were passed locally.
```
<class name="org.wso2.identity.integration.test.rest.api.server.idp.v1.IdPSuccessTest"/>
<class name="org.wso2.identity.integration.test.rest.api.server.idp.v1.IdPFailureTest"/>
```

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 1]
```diff
@@ -2772,4 +2775,119 @@ private static String includeData(Constants.ErrorMessage error, String data) {
...
+     *                                             config is found.
+     */
+    private void validateClaims(String tenantDomain, Claims claims) throws IdentityProviderManagementException {
+
+        String userClaimURI = claims.getUserIdClaim().getUri();
+        String roleClaimURI = claims.getRoleClaim().getUri();
+        List<org.wso2.carbon.identity.api.server.idp.v1.model.ClaimMapping> claimMappings = claims.getMappings();
+
+        // EMPTY claimMappings indicate that the IDP is using local claim dialect.
+        if (claimMappings == null || claimMappings.isEmpty()) {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2794]: Can use CollectionUtils.isEmpty

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 2]
```diff
@@ -2772,4 +2775,119 @@ private static String includeData(Constants.ErrorMessage error, String data) {
...
+     *                                                   the claim mappings.
+     */
+    private void validateUserAndRoleClaims(String userClaimURI, String roleClaimURI,
+                                List<org.wso2.carbon.identity.api.server.idp.v1.model.ClaimMapping> claimMappings)
+            throws IdentityProviderManagementClientException {
+
+        boolean isValidUserClaimURI = false;
+        boolean isValidRoleClaimURI = false;
+        for (org.wso2.carbon.identity.api.server.idp.v1.model.ClaimMapping claimMapping : claimMappings) {
+            if (userClaimURI.equals(claimMapping.getIdpClaim())) {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2844]: can `userClaimURI` or `roleClaimURI` be null?

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 3]
```diff
@@ -2772,4 +2775,119 @@ private static String includeData(Constants.ErrorMessage error, String data) {
...
+
+        try {
+            List<LocalClaim> localClaimsList =
+                    IdentityProviderServiceHolder.getClaimMetadataManagementService().getLocalClaims(tenantDomain);
+            if (localClaimsList.isEmpty()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("No local claims found for tenant:" + tenantDomain + ".Therefore, skipping " +
+                            "local claim URI validation.");
+                }
+                return new ArrayList<>();
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2881]: Use the util method Collections.emptyList() instead

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 4]
```diff
@@ -2772,4 +2776,125 @@ private static String includeData(Constants.ErrorMessage error, String data) {
...
+        String userClaimURI = claims.getUserIdClaim() == null ? null : claims.getUserIdClaim().getUri();
+        String roleClaimURI = claims.getRoleClaim() == null ? null : claims.getRoleClaim().getUri();
+        List<org.wso2.carbon.identity.api.server.idp.v1.model.ClaimMapping> claimMappings = claims.getMappings();
+
+        // EMPTY claimMappings indicate that the IDP is using local claim dialect.
+        if (CollectionUtils.isEmpty(claimMappings)) {
+            List<LocalClaim> localClaimsList = getLocalClaimURIs(tenantDomain);
+            Set<String> claimURIs = localClaimsList.stream().map(LocalClaim::getClaimURI).collect(Collectors.toSet());
+            // Validate userClaimURI and roleClaimURI.
+            if (StringUtils.isNotBlank(userClaimURI) && claimURIs.contains(userClaimURI)) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2802]: Shoudn't this be not contains? 

!claimURIs.contains(userClaimURI)

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 5]
```diff
@@ -2772,4 +2776,125 @@ private static String includeData(Constants.ErrorMessage error, String data) {
...
+        }
+        validateUserAndRoleClaims(userClaimURI, roleClaimURI, claimMappings);
+
+        // Validate LocalClaim objects against local claim URIs.
+        List<LocalClaim> localClaimsList = getLocalClaimURIs(tenantDomain);
+        Set<String> claimURIs = localClaimsList.stream().map(LocalClaim::getClaimURI).collect(Collectors.toSet());
+        for (org.wso2.carbon.identity.api.server.idp.v1.model.ClaimMapping claimMapping : claimMappings) {
+
+            // If a claim URI can be added to the existing claimURIs, then that's a not existing URI.
+            if (claimURIs.contains(claimMapping.getLocalClaim().getUri())) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2824]: Shoud be not contain? 
            if (!claimURIs.contains(claimMapping.getLocalClaim().getUri())) {

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java [hunk 6]
```diff
@@ -116,6 +116,16 @@
         ERROR_CODE_OUTBOUND_PROVISIONING_CONFIG_NOT_FOUND("60029", "Unable to update Outbound " +
                 "Provisioning Connector", "Outbound Provisioning Connector properties have not specified " +
                 "for connector : %s"),
+        ERROR_CODE_INVALID_USER_CLAIM_URI("60030", "Invalid user ID claim URI",
+                "User ID claim URI: %s does not match with the claim mappings"),
+        ERROR_CODE_INVALID_ROLE_CLAIM_URI("60031", "Invalid role claim URI",
+                "Role claim URI: %s does not match with the claim mappings"),
+        ERROR_CODE_NOT_EXISTING_CLAIM_URI("IDP-60032", "Invalid claim URI",
+                "One or more local claim URIs does not exist"),
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java:124]: grammer :  URIs does not exist

Comment (somindatommy) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java]: Fixed in f44b32e0efda15349b0bbf3d014763ee0b8297ad

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 5 above)
Comment (somindatommy) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Fixed in f44b32e0efda15349b0bbf3d014763ee0b8297ad

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 4 above)
Comment (somindatommy) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Fixed in f44b32e0efda15349b0bbf3d014763ee0b8297ad

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 7]
```diff
@@ -2842,26 +2842,32 @@ private void validateUserAndRoleClaims(String userClaimURI, String roleClaimURI,
...
 
-        boolean isUserClaimURISpecified = StringUtils.isNotBlank(userClaimURI);
-        boolean isRoleClaimURISpecified = StringUtils.isNotBlank(roleClaimURI);
-        boolean isValidUserClaimURI = false;
-        boolean isValidRoleClaimURI = false;
+        /*
+         * Not defining the userId claim or role claim is a valid claim update scenario. Therefore, considering it as a
+         * valid claim URI. When the claims are defined, initial validation is considered as false (Invalid).
+         */
+        boolean isValidUserClaimURI = StringUtils.isBlank(userClaimURI);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2849]: Shall we rename to isUserClaimURINotDefined, isRoleClaimURINotDefined

Comment (IsuraD) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2849]: or. better to use 

```
boolean isUserClaimURIDefined = StringUtils.isNotBlank(userClaimURI);
boolean isRoleClaimURIDefined = StringUtils.isNotBlank(roleClaimURI);
```

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java [hunk 8]
```diff
@@ -116,6 +116,16 @@
         ERROR_CODE_OUTBOUND_PROVISIONING_CONFIG_NOT_FOUND("60029", "Unable to update Outbound " +
                 "Provisioning Connector", "Outbound Provisioning Connector properties have not specified " +
                 "for connector : %s"),
+        ERROR_CODE_INVALID_USER_CLAIM_URI("60030", "Invalid user ID claim URI",
+                "User ID claim URI: %s does not match with the claim mappings"),
+        ERROR_CODE_INVALID_ROLE_CLAIM_URI("60031", "Invalid role claim URI",
+                "Role claim URI: %s does not match with the claim mappings"),
+        ERROR_CODE_NOT_EXISTING_CLAIM_URI("IDP-60032", "Invalid claim URI",
```
[the remark this answers is not in the document]
Comment (somindatommy) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java:123]: Fixed in https://github.com/wso2/identity-api-server/pull/216/commits/0d758b5bfff01a26c5caa8d68cf92ad084d86927

[the remark this answers is not in the document]
Comment (somindatommy) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java:123]: Ideally, they should have the code. AFAIK at the time of the initial implementation, the scenario prefix concept was not there.

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java [hunk 9]
```diff
@@ -213,7 +223,10 @@
                 "Error occurred while updating the IDP template with identifier %s."),
         ERROR_CODE_ERROR_RETRIEVING_IDP_TEMPLATE("65054", "Unable to retrieve IDP template.",
                 "Error occurred while retrieving the IDP template with identifier %s"),
-        ERROR_CODE_ERROR_INVALID_SEARCH_FILTER("65055", "Search request validation failed.", "Invalid search filter.");
+        ERROR_CODE_ERROR_INVALID_SEARCH_FILTER("65055", "Search request validation failed.",
+                "Invalid search filter."),
+        ERROR_CODE_VALIDATING_LOCAL_CLAIM_URIS("IDP-65056", "Error while validation local claim URIs",
```
[the remark this answers is not in the document]
Comment (somindatommy) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java:228]: Addressed in the previous comment. I have not specified the existing code for `ERROR_CODE_ERROR_INVALID_SEARCH_FILTER`

**Review by IsuraD — APPROVED** [review]

**Pull request merged**